src/transforms/rename_fields.rs

```diff
@@ -44,12 +44,19 @@ impl TransformConfig for RenameFieldsConfig {
 
 impl RenameFields {
     pub fn new(fields: IndexMap<String, TomlValue>) -> crate::Result<Self> {
-        Ok(RenameFields {
-            fields: fields
-                .into_iter()
-                .map(|kv| flatten(kv, None))
-                .collect::<crate::Result<_>>()?,
-        })
+        let fields = fields
+            .into_iter()
+            .map(|kv| flatten(kv, None))
+            .collect::<crate::Result<Vec<_>>>()?;
+        let fields = fields
+            .into_iter()
+            .flat_map(|v| v)
+            .map(|v| match v {
+                (origin, TomlValue::String(new)) => Ok((origin.into(), new.into())),
+                (origin, _) => Err(FlattenError::CannotFlatten { key: origin }.into()),
```

Comment:
Could the error handling be offloaded on Serde with assistance of the type system?

For example, it might be possible to set the type of `fields` in the config struct to a newly defined `Fields` type:

```rust
#[derive(Deserialize)]
struct Fields(IndexMap<String, FieldsOrValue<String>>);

#[derive(Deserialize)]
#[serde(untagged)]
enum FieldsOrValue {
    Fields(Fields),
    Value(String),
}
```

And implement `all_fields` method for `Fields` which would return a list of flattened fields and their corresponding values, without the need to check for errors.

Also, could this part of code be shared between `rename_fields` and `add_fields`, for example by placing it into a separate file such in `src/transforms/util`? In that case further changes such as adding support for escaping in the field path notation would require to change only a single place of code.

Here is a full example of the auxiliary types approach which is generic over the target value type: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9feba91b1777279144dc6c0630b13c74.

Reply:
I think that's a great idea, Alex!